## 텍스트의 토큰화
**자연어 처리(NLP)에서 토큰화(Tokenization)**
- 텍스트 분석, 언어 번역, 감정 분석 등 다양한 NLP 작업을 위해 텍스트를 준비하는 기본 단계입니다.
- 토큰화는 텍스트를 단어, 문구, 기호 또는 기타 의미 있는 요소(토큰)로 분해하는 작업을 포함합니다. 이 과정에서 생성된 토큰은 추가 처리 및 분석을 위한 기본 구성 요소가 됩니다.

토큰화의 목적
- 토큰화의 주요 목적은 텍스트 데이터를 단순화하여 알고리즘이 이해하고 처리할 수 있도록 관리하기 쉽게 만드는 것입니다. 이를 통해 텍스트의 복잡성을 줄이고 일관성을 유지함으로써, 다양한 NLP 작업에서 효율적인 분석과 처리가 가능해집니다.

토큰화의 유형
- 토큰화는 다양한 수준에서 수행될 수 있으며, 각 유형은 특정 NLP 작업의 요구 사항에 따라 선택됩니다:
  - 단어 토큰화 (Word Tokenization):
    텍스트를 개별 단어로 분해합니다.
    예: "ChatGPT is amazing!" → ["ChatGPT", "is", "amazing", "!"]
  - 문장 토큰화 (Sentence Tokenization):
    텍스트를 개별 문장으로 분해합니다.
    예: "Hello world. How are you?" → ["Hello world.", "How are you?"]
  - 하위 단어 토큰화 (Subword Tokenization):
    단어를 더 작은 의미 단위로 분해합니다. 주로 BPE(Byte Pair Encoding)나 WordPiece 알고리즘을 사용합니다.
    예: "unhappiness" → ["un", "hap", "pi", "ness"]
  - 문자 토큰화 (Character Tokenization):
    텍스트를 개별 문자로 분해합니다.
    예: "Hello" → ["H", "e", "l", "l", "o"]

토큰화의 과정
- 토큰화 과정은 일반적으로 다음 단계로 구성됩니다:
  - 텍스트 정규화 (Text Normalization):
    모든 텍스트를 소문자로 변환하여 일관성을 유지합니다.
    불필요한 구두점과 공백을 제거합니다.
    예: "Hello, World!" → "hello world"
  - 구분자 사용 (Delimiter-based Tokenization):
    공백이나 구두점을 기준으로 텍스트를 분해합니다.
    예: "hello world" → ["hello", "world"]
  - 고급 토큰화 기법 (Advanced Tokenization Techniques):
    언어의 문법적, 의미적 구조를 고려하여 토큰을 생성합니다.
    BPE, WordPiece, SentencePiece 등의 알고리즘을 사용합니다.

토큰화의 중요성
- 토큰화는 NLP 작업에서 매우 중요한 역할을 합니다. 잘못된 토큰화는 후속 처리와 분석의 정확도에 큰 영향을 미칠 수 있습니다.
- 반면, 올바른 토큰화는 텍스트 데이터를 효과적으로 전처리하고 분석할 수 있게 합니다.

## 고급 토큰화 기법
언어의 문법적, 의미적 구조를 고려하여 토큰을 생성하는 것은 텍스트의 의미를 더 잘 보존하고, 더 정확한 분석과 처리를 가능하게 하기 위한 고급 토큰화 기법입니다. 이러한 기법들은 단순히 공백이나 구두점을 기준으로 텍스트를 분해하는 것을 넘어, 단어의 의미와 형태, 문장의 구조 등을 이해하여 더 정교한 토큰을 생성합니다.


형태소 분석 (Morphological Analysis):
- 단어를 구성하는 최소 의미 단위인 형태소를 분석합니다.
- 예를 들어, "cats"는 "cat"과 복수형 접미사 "s"로 분해됩니다.
- 형태소 분석기는 단어의 어간과 접사(접두사, 접미사)를 인식하고 분리합니다.

어간 추출 (Stemming):
- 단어의 어간을 추출하여 형태를 단순화합니다.
- 예: "running", "runs", "ran" → "run"
- 포터 스테머(Porter Stemmer)와 같은 알고리즘이 사용됩니다.

어근 추출 (Lemmatization):
- 단어의 어근을 추출하여 형태를 표준화합니다. 어간 추출보다 더 정교합니다.
- 예: "running", "ran" → "run"
- 품사 정보를 사용하여 정확한 어근을 찾아냅니다. 예를 들어, "better"는 어근 "good"으로 변환됩니다.
- WordNetLemmatizer와 같은 도구가 사용됩니다.

BPE (Byte Pair Encoding):
- 자주 등장하는 바이트 쌍을 병합하여 점진적으로 단어를 분해합니다.
- 예: "lowest"가 "l", "o", "w", "e", "s", "t"로 분해되고, 자주 등장하는 "lo", "we"가 결합되어 "low", "est"로 변환됩니다.
- BPE는 신경망 번역 모델과 같은 대규모 언어 모델에서 널리 사용됩니다.

WordPiece:
- BPE와 유사하지만, 서브워드(subword) 단위로 토큰을 생성합니다.
- 예: "unhappiness" → ["un", "##happiness"]
- 트랜스포머 모델(BERT 등)에서 사용됩니다.

SentencePiece:
- 언어에 중립적인 방식으로 텍스트를 서브워드 단위로 분해합니다.
- BPE와 유사하지만, 문장을 토큰화하는 과정에서 공백을 고려하지 않음.
- 예: "unhappiness" → ["un", "ha", "ppiness"]
- Google의 T5, ALBERT 모델에서 사용됩니다.

In [ ]:
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten, Embedding
from tensorflow.keras.utils import to_categorical
from numpy import array

# 텍스트 전처리와 관련한 함수 중  text_to_word_sequence를 부러오기
from tensorflow.keras.preprocessing.text import text_to_word_sequence

text = '해보지 않으면 해낼 수 없다'

result = text_to_word_sequence(text)
print('\n원문: \n',text)
print('\n토큰화:\n',result)


원문: 
 해보지 않으면 해낼 수 없다

토큰화:
 ['해보지', '않으면', '해낼', '수', '없다']


- Tokenizer' 클래스는 텍스트를 정수 시퀀스로 변환하도록 설계
- fit_on_texts(docs): 이 메소드는 문장 목록(docs)을 인수로 사용하여 token 개체에서 호출된다. 텍스트 목록을 기반으로 내부 어휘를 업데이트하여 토크나이저가 이러한 텍스트로 작업할 수 있도록 준비한다. 말뭉치의 각 고유 단어에 색인을 할당하고 단어 빈도와 같은 다양한 측정항목을 계산하는 작업이 포함된다.
- token.word_counts: 토크나이저를 텍스트에 맞춘 후 word_counts는 키가 입력 텍스트에서 발견된 단어이고 값은 각 단어의 발생 횟수인 OrderedDict를 제공한다. 'OrderedDict'를 사용하면 단어가 텍스트에서 처음 나타나는 순서대로 정렬.
- token.document_count: 이 속성은 처리된 총 문서(또는 문장) 수를 표시
- token.word_docs: word_counts와 유사한 OrderedDict이지만 단어의 빈도 대신 각 단어가 나타나는 문서 수를 표시
- token.word_index: 이 속성은 단어를 고유하게 할당된 정수에 매핑하는 OrderedDict를 제공. 모델에는 숫자 입력이 필요하므로 이는 기계 학습 모델의 텍스트를 벡터화하는 데 필요

In [ ]:
docs = ['먼저 텍스트의 각 단어를 나누어 토큰화합니다.',
        '텍스트의 단어로 토큰화해야 딥러닝에서 인식됩니다.',
        '토큰화한 결과는 딥러닝에서 사용할 수 있습니다.']

token = Tokenizer() #토큰화 함수 지정
token.fit_on_texts(docs) #토큰화 함수에 문장 적용

# 단어의 빈도수를 계산한 결과를 각 옵션에 맞추어 출력
# Tokenizer()의 word_counts함수는 순서를 기억하는 OrderedDict 클래스를 사용
print('\n단어 카운트:\n', token.word_counts)
print('\n문장 카운트:',token.document_count)
print('\n각 단어가 몇 개의 문장에 포함되어 있는가:\n', token.word_docs)
# token.word_index의 출력 순서는 제공된 텍스트 코퍼스의 각 단어의 빈도에 따라 가장 빈번한 단어부터 가장 빈도가 낮은 단어까지 결정
# 동일한 빈도의 경우는 먼저 등장한 단어가 더 낮은 인덱스를 할당
print('\n각 단어에 매겨진 인덱스 값:\n', token.word_index)


단어 카운트:
 OrderedDict([('먼저', 1), ('텍스트의', 2), ('각', 1), ('단어를', 1), ('나눈어', 1), ('토큰화합니다', 1), ('단어로', 1), ('토큰화해야', 1), ('딥러닝에서', 2), ('인식됩니다', 1), ('토큰화한', 1), ('결과는', 1), ('사용할', 1), ('수', 1), ('있습니다', 1)])

문장 카운트: 3

각 단어가 몇 개의 문장에 포함되어 있는가:
 defaultdict(<class 'int'>, {'단어를': 1, '각': 1, '토큰화합니다': 1, '먼저': 1, '나눈어': 1, '텍스트의': 2, '단어로': 1, '인식됩니다': 1, '토큰화해야': 1, '딥러닝에서': 2, '토큰화한': 1, '있습니다': 1, '결과는': 1, '사용할': 1, '수': 1})

각 단어에 매겨진 인덱스 값:
 {'텍스트의': 1, '딥러닝에서': 2, '먼저': 3, '각': 4, '단어를': 5, '나눈어': 6, '토큰화합니다': 7, '단어로': 8, '토큰화해야': 9, '인식됩니다': 10, '토큰화한': 11, '결과는': 12, '사용할': 13, '수': 14, '있습니다': 15}


In [ ]:
docs = ['검찰이 제시한 혐의 사실 전부를 재판부가 무죄로 판단하면서 이 회장은 검찰 기소 이후 3년 5개월여 만에 시름을 덜게 됐다.',
'검찰 항소로 2심 재판이 진행될 것이란 전망이 나오지만,']

token = Tokenizer()
token.fit_on_texts(docs)

print('\n단어 카운트:\n', token.word_counts)
print('\n문장 카운트:',token.document_count)
print('\n각 단어가 몇개의 문장에 포함되어있는가:\n', token.word_docs)
print('\n각 단어에 매겨진 인덱스 값:\n', token.word_index)


단어 카운트:
 OrderedDict([('검찰이', 1), ('제시한', 1), ('혐의', 1), ('사실', 1), ('전부를', 1), ('재판부가', 1), ('무죄로', 1), ('판단하면서', 1), ('이', 1), ('회장은', 1), ('검찰', 2), ('기소', 1), ('이후', 1), ('3년', 1), ('5개월여', 1), ('만에', 1), ('시름을', 1), ('덜게', 1), ('됐다', 1), ('항소로', 1), ('2심', 1), ('재판이', 1), ('진행될', 1), ('것이란', 1), ('전망이', 1), ('나오지만', 1)])

문장 카운트: 2

각 단어가 몇개의 문장에 포함되어있는가:
 defaultdict(<class 'int'>, {'만에': 1, '재판부가': 1, '전부를': 1, '판단하면서': 1, '시름을': 1, '혐의': 1, '3년': 1, '됐다': 1, '기소': 1, '덜게': 1, '5개월여': 1, '사실': 1, '이': 1, '무죄로': 1, '검찰이': 1, '회장은': 1, '제시한': 1, '검찰': 2, '이후': 1, '2심': 1, '진행될': 1, '전망이': 1, '재판이': 1, '항소로': 1, '것이란': 1, '나오지만': 1})

각 단어에 매겨진 인덱스 값:
 {'검찰': 1, '검찰이': 2, '제시한': 3, '혐의': 4, '사실': 5, '전부를': 6, '재판부가': 7, '무죄로': 8, '판단하면서': 9, '이': 10, '회장은': 11, '기소': 12, '이후': 13, '3년': 14, '5개월여': 15, '만에': 16, '시름을': 17, '덜게': 18, '됐다': 19, '항소로': 20, '2심': 21, '재판이': 22, '진행될': 23, '것이란': 24, '전망이': 25, '나오지만': 26}


In [ ]:
import nltk
nltk.download('punkt')


[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

In [ ]:
import nltk
nltk.download('wordnet')


[nltk_data] Downloading package wordnet to /root/nltk_data...


True

In [ ]:
from nltk.stem import PorterStemmer
from nltk.stem import WordNetLemmatizer
from nltk.tokenize import word_tokenize

# 어간 추출기와 표제어 추출기 초기화
stemmer = PorterStemmer()
lemmatizer = WordNetLemmatizer()

# 예제 문장
sentence = "과일을 과일이 과일이라고"
tokens = word_tokenize(sentence)

# 어간 추출
stems = [stemmer.stem(token) for token in tokens]
print("어간 추출:", stems)

# 표제어 추출
lemmas = [lemmatizer.lemmatize(token) for token in tokens]
print("표제어 추출:", lemmas)


어간 추출: ['과일을', '과일이', '과일이라고']
표제어 추출: ['과일을', '과일이', '과일이라고']


In [ ]:
!pip install konlpy


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 19.4/19.4 MB 73.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 488.6/488.6 kB 29.5 MB/s eta 0:00:00


In [ ]:
from konlpy.tag import Okt

# Okt 형태소 분석기 초기화
okt = Okt()

# 예제 문장
sentence = "과일을 과일이 과일이라고"

# 형태소 분석 및 표제어 추출
tokens = okt.morphs(sentence)
print("형태소 분석:", tokens)

# 표제어 추출
# 한국어에서 표제어 추출을 위해서는 형태소 분석을 통해 얻은 결과를 활용
# '과일'이 동일한 형태로 유지되지만, 일반적으로 '형태소'가 추출됨.


형태소 분석: ['과일', '을', '과일', '이', '과일', '이라고']


In [ ]:
from konlpy.tag import Okt
from collections import Counter

# Okt 형태소 분석기 초기화
okt = Okt()

# 예제 문장
sentence = "과일을 과일이 과일이라고"

# 형태소 분석
tokens = okt.morphs(sentence)
print("형태소 분석 결과:", tokens)

# 형태소의 빈도 계산
token_counts = Counter(tokens)
print("형태소 빈도:", token_counts)


형태소 분석 결과: ['과일', '을', '과일', '이', '과일', '이라고']
형태소 빈도: Counter({'과일': 3, '을': 1, '이': 1, '이라고': 1})


Keras의 Tokenizer를 사용하여 텍스트 데이터를 정수 인덱스 시퀀스로 변환한 후, 이를 One-Hot Encoding 형식으로 변환하는 과정은 NLP 모델의 입력 데이터를 준비하는 중요한 단계입니다.

Tokenizer를 사용한 텍스트 토큰화

word_index:
- token.word_index는 각 단어를 고유한 정수 인덱스로 매핑한 딕셔너리입니다. 키는 단어이고, 값은 해당 단어의 인덱스입니다.
- 이 딕셔너리의 길이(len(token.word_index))는 말뭉치에 있는 고유 단어의 총 개수를 나타냅니다.

word_size:
- word_size는 고유 단어의 총 개수에 1을 더한 값입니다. 이는 NLP에서 일반적인 관행으로, "0" 인덱스를 포함하기 위해 사용됩니다.
- "0" 인덱스는 패딩(padding)에 사용되거나, 구현에 따라 알 수 없는 단어를 나타낼 수 있습니다.

One-Hot Encoding:
- to_categorical 함수는 클래스 벡터(정수 인덱스)를 바이너리 클래스 행렬로 변환합니다.
- x는 단어를 나타내는 정수 인덱스의 목록 또는 배열이어야 하며, - to_categorical은 이를 One-Hot Encoding 형식으로 변환합니다.
- 각 정수에 대해 해당 인덱스 위치만 1로 설정되고 나머지 위치는 0인 벡터를 생성합니다.

num_classes:
- num_classes는 총 클래스 수를 지정합니다. 이 경우 어휘 크기(word_size)로 설정되어, One-Hot Encoding에 어휘의 모든 단어에 대한 슬롯과 추가 "0" 인덱스가 있는지 확인합니다.

Embedding 레이어
- 자연어 처리나 시퀀스 데이터를 다루는 모델에서 자주 사용되는 레이어로, 입력 데이터를 고정된 크기의 고차원 벡터로 변환해주는 역할을 합니다.

주요 기능:
  - 단어를 벡터로 변환: Embedding 레이어는 고유한 정수 인덱스로 표현된 단어들을 특정 크기의 벡터로 변환합니다. 이 벡터들은 학습 과정에서 모델이 조정할 수 있는 가중치로 초기화됩니다.
  - 차원 축소 및 특성 학습: 높은 차원의 단어를 저차원의 벡터로 변환함으로써 데이터의 차원을 축소하고, 단어 간의 유사성을 학습할 수 있습니다. 예를 들어, 비슷한 의미의 단어들이 유사한 벡터로 매핑되도록 학습됩니다.

입력 및 출력:
  - 입력: 정수로 인코딩된 시퀀스 데이터. 예를 들어, 단어를 고유한 정수로 인코딩한 시퀀스.
  - 출력: 각 입력 정수에 대응하는 고정된 크기의 밀집 벡터.

In [ ]:
# keras Tokenizer의 fit_on_texts 메소드는 일반적으로 문자열의 리스트를 기대하기 때문에, 단일 문자열을 입력하는 것은 적절하지 않다
text = '오랫동안 꿈꾸는 이는 그 꿈을 닮아간다'
token = Tokenizer()
token.fit_on_texts([text])
print(token.word_index)

{'오랫동안': 1, '꿈꾸는': 2, '이는': 3, '그': 4, '꿈을': 5, '닮아간다': 6}


In [ ]:
text = '오랫동안 꿈꾸는 이는 그 꿈을 닮아간다'
token = Tokenizer()
token.fit_on_texts(text)
print(token.word_index)

{'꿈': 1, '는': 2, '오': 3, '랫': 4, '동': 5, '안': 6, '꾸': 7, '이': 8, '그': 9, '을': 10, '닮': 11, '아': 12, '간': 13, '다': 14}


In [ ]:
x = token.texts_to_sequences([text])
print(x)

[[1, 2, 3, 4, 5, 6]]


In [ ]:
# 인덱스 수에 하나를 추가해서 원-핫 인코딩 배열 만들기
# word_size에 1을 추가하는 이유는 Keras의 Tokenizer를 사용할 때 0 인덱스를 패딩을 위해 예약하는 관례 때문
word_size = len(token.word_index) +1
x = to_categorical(x, num_classes=word_size)
print(x)

[[[0. 1. 0. 0. 0. 0. 0.]
  [0. 0. 1. 0. 0. 0. 0.]
  [0. 0. 0. 1. 0. 0. 0.]
  [0. 0. 0. 0. 1. 0. 0.]
  [0. 0. 0. 0. 0. 1. 0.]
  [0. 0. 0. 0. 0. 0. 1.]]]


In [2]:
import tensorflow as tf
from tensorflow.keras.layers import Embedding, Input

# 입력 시퀀스의 길이를 3이라고 가정
input_data = tf.constant([1,2,3], dtype=tf.int32)
# 임베딩 레이어 정의
embedding_layer = Embedding(input_dim=100, output_dim=10)
# 텐서로 전달해야함
output = embedding_layer(input_data)

print(output)

tf.Tensor(
[[-3.7558388e-02  3.2512639e-02  2.5352802e-02  2.1997143e-02
   4.2947087e-02 -4.8430171e-02 -1.9364893e-02 -2.3285603e-02
  -2.5157213e-02  3.2960925e-02]
 [-2.8330907e-03 -1.0687493e-02  3.7445772e-02  1.5826408e-02
   2.1206427e-02  2.3162965e-02  7.6342113e-03 -2.8181219e-02
  -2.1670057e-02 -3.9850365e-02]
 [ 9.0622194e-03  4.3486405e-02 -3.7891507e-02 -4.7203567e-02
  -8.9071691e-05 -1.7407648e-03 -7.1063414e-03  4.1811708e-02
  -1.5952073e-02 -1.5902296e-03]], shape=(3, 10), dtype=float32)


In [3]:
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten, Embedding
from tensorflow.keras.utils import to_categorical
from numpy import array



In [4]:
docs = ['너무 재밌네요','최고예요','참 잘 만든 영화예요','추천하고 싶은 영화입니다','한번 더 보고싶네요','글쎄요','별로예요','생각보다 지루하네요','연기가 어색해요','재미없어요']

classes = array([1,1,1,1,1,0,0,0,0,0])

token = Tokenizer()
token.fit_on_texts(docs)
print(token.word_index)

{'너무': 1, '재밌네요': 2, '최고예요': 3, '참': 4, '잘': 5, '만든': 6, '영화예요': 7, '추천하고': 8, '싶은': 9, '영화입니다': 10, '한번': 11, '더': 12, '보고싶네요': 13, '글쎄요': 14, '별로예요': 15, '생각보다': 16, '지루하네요': 17, '연기가': 18, '어색해요': 19, '재미없어요': 20}


In [5]:
x = token.texts_to_sequences(docs)
print('\n리뷰 텍스트, 토큰화 결과:\n', x)


리뷰 텍스트, 토큰화 결과:
 [[1, 2], [3], [4, 5, 6, 7], [8, 9, 10], [11, 12, 13], [14], [15], [16, 17], [18, 19], [20]]


In [7]:
# 패딩, 서로 다른 길이의 데이터를 4로 맞추어 줍니다
# padding = 'pre' 시퀀스의 길이가 4보다 짧은 경우 앞쪽을 0으로 채워 길이를 4로 맞춘다
padded_x = pad_sequences(x,4,padding='pre')
# pre앞에 채워짐 post 뒤에 채워짐

print('\n패딩 결과:\n',padded_x)


패딩 결과:
 [[ 0  0  1  2]
 [ 0  0  0  3]
 [ 4  5  6  7]
 [ 0  8  9 10]
 [ 0 11 12 13]
 [ 0  0  0 14]
 [ 0  0  0 15]
 [ 0  0 16 17]
 [ 0  0 18 19]
 [ 0  0  0 20]]


### Embedding
- 기계 학습, 특히 자연어 처리(NLP)의 맥락에서 임베딩은 단어, 구문 또는 기타 유형의 엔터티를 실수의 밀집된 벡터로 표현하는 기술을 의미. 핵심 아이디어는 이러한 엔터티의 의미론적 의미를 연속적인 벡터 공간으로 인코딩하는 것이다. 여기서 벡터 간의 기하학적 관계는 이들이 나타내는 엔터티 간의 의미론적 관계를 반영하며 이 접근 방식은 의미론적 관계가 캡처되지 않는 원-핫 인코딩과 같은 희소 표현과 대조된다.
- 의미:
  - 의미적 표현: 임베딩 벡터는 단어나 개체의 의미를 포착하도록 설계된다. 비슷한 의미를 가진 단어는 임베딩 공간에서 서로 가까운 벡터를 갖도록 처리된다.
  - 차원성 감소: 임베딩은 고차원 공간(예: 원-핫 인코딩된 벡터)의 단어를 저차원의 연속 벡터 공간으로 매핑하며 이는 표현을 더욱 효율적으로 만들고 계산 복잡성을 줄인다.
  - 컨텍스트화: 고급 임베딩 모델(예: Word2Vec, GloVe, BERT)에서는 단어가 나타나는 컨텍스트가 벡터 표현에 영향을 미치므로 모델이 다양한 컨텍스트에서 단어의 다양한 의미를 캡처할 수 있다.
- 임베딩 생성 방법:
  - 사전 훈련된 임베딩: 일반적인 접근 방식 중 하나는 대규모 텍스트 모음에 대해 사전 훈련된 임베딩을 사용하는 것으로 Word2Vec, GloVe와 같은 모델을 사용하면 훈련 코퍼스에서 풍부한 의미 체계 관계를 학습한 사전 훈련된 모델을 기반으로 단어를 벡터에 매핑할 수 있다.
    - Word2Vec: 컨텍스트를 사용하여 대상 단어를 예측하거나(CBOW) 단어를 사용하여 컨텍스트를 예측(Skip-gram)하여 임베딩을 학습한다.
      - CBOW 모델은 주어진 컨텍스트(주변 단어들)를 바탕으로 타겟 단어를 예측하는 방식으로 작동. 예를 들어, 문장 "the cat sits on the"에서 "cat", "sits", "on", "the"를 컨텍스트로 사용하여 "mat"이라는 타겟 단어를 예측
      - skip-gram 모델은 CBOW와 반대로, 주어진 타겟 단어로부터 컨텍스트(주변 단어들)를 예측하는 방식으로 작동. 예를 들어, "cat"이라는 단어가 주어졌을 때, "the", "sits", "on", "the"와 같은 주변 단어들을 예측
    - GloVe: 단어 동시 발생 통계 행렬을 인수분해하여 임베딩을 학습한다.    
  - 자신만의 임베딩 훈련: 신경망을 사용하여 처음부터 자신만의 임베딩을 훈련할 수도 있다.

텍스트 입력을 표현하기 위해 단어 임베딩 활용
- model.add(Embedding(word_size, 8, input_length=4)): 이 줄은 모델에 임베딩 레이어를 추가. 임베딩 레이어는 단어 임베딩을 만드는 데 사용. 각 매개변수의 의미는 다음과 같다.
  - word_size: 입력 차원의 크기, 즉 어휘 크기. 데이터 세트에 있는 총 고유 단어 수에 1을 더한 값.
  - 8: 임베딩 벡터의 크기. 각 단어는 8차원 벡터로 표현.
  - input_length=4: 입력 시퀀스의 길이. 이 모델은 각 입력 시퀀스의 길이가 4일 것으로 예상(예: 입력당 4개의 단어).
- model.add(Flatten()): 임베딩 레이어 이후 출력 모양은 배치 크기를 포함하여 3차원. 이 레이어는 출력을 2차원(배치 크기, input_length * 8)으로 평면화하여 밀도가 높은 레이어에 직접 공급될 수 있도록 한다.

In [8]:
import numpy as np

docs = ['너무 재밌네요','최고예요','참 잘 만든 영화예요','추천하고 싶은 영화입니다','한번 더 보고싶네요','글쎄요','별로예요','생각보다 지루하네요','연기가 어색해요','재미없어요']

classes = np.array([1,1,1,1,1,0,0,0,0,0])


In [10]:
from tensorflow.keras.preprocessing.text import Tokenizer

token = Tokenizer()
token.fit_on_texts(docs)

print(token.word_index)

word_size = len(token.word_index) +1

model = Sequential()
model.add(Input(shape=(4,)))
model.add(Embedding(word_size, 8))
model.add(Flatten())
model.add(Dense(1, activation='sigmoid'))
model.summary()

{'너무': 1, '재밌네요': 2, '최고예요': 3, '참': 4, '잘': 5, '만든': 6, '영화예요': 7, '추천하고': 8, '싶은': 9, '영화입니다': 10, '한번': 11, '더': 12, '보고싶네요': 13, '글쎄요': 14, '별로예요': 15, '생각보다': 16, '지루하네요': 17, '연기가': 18, '어색해요': 19, '재미없어요': 20}


Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ embedding_1 (Embedding)              │ (None, 4, 8)                │             168 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ flatten (Flatten)                    │ (None, 32)                  │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense (Dense)                        │ (None, 1)                   │              33 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 201 (804.00 B)

 Trainable params: 201 (804.00 B)

 Non-trainable params: 0 (0.00 B)

In [11]:
x = token.texts_to_sequences(docs)
padded_x = pad_sequences(x, 4, padding='pre')

In [13]:
model.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])
model.fit(padded_x, classes, epochs=20)
print('\n Accuracy: %.4f' % (model.evaluate(padded_x, classes)[1]))

Epoch 1/20
1/1 ━━━━━━━━━━━━━━━━━━━━ 1s 856ms/step - accuracy: 0.9000 - loss: 0.6517
Epoch 2/20
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 34ms/step - accuracy: 0.9000 - loss: 0.6497
Epoch 3/20
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 55ms/step - accuracy: 0.9000 - loss: 0.6478
Epoch 4/20
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 31ms/step - accuracy: 0.9000 - loss: 0.6458
Epoch 5/20
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 30ms/step - accuracy: 0.9000 - loss: 0.6438
Epoch 6/20
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 29ms/step - accuracy: 0.9000 - loss: 0.6418
Epoch 7/20
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 29ms/step - accuracy: 0.9000 - loss: 0.6397
Epoch 8/20
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 35ms/step - accuracy: 0.9000 - loss: 0.6377
Epoch 9/20
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 41ms/step - accuracy: 0.9000 - loss: 0.6357
Epoch 10/20
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 58ms/step - accuracy: 0.9000 - loss: 0.6336
Epoch 11/20
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 56ms/step - accuracy: 0.9000 - loss: 0.6316
Epoch 12/20
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 53ms/step - accuracy: 0.9000 - loss: 0.6295


Q. 10개의 단어를 가진 어휘(vocabulary)와 각 단어를 4차원 벡터로 임베딩하는 Keras 모델을 구성하세요. 시퀀스의 최대 길이는 5로 설정하세요.

`input_data = np.array([
    [1, 2, 3, 4, 5],   
    [2, 3, 4, 5, 6]    
])`

In [14]:
input_data = np.array([ [1, 2, 3, 4, 5], [2, 3, 4, 5, 6] ])

model = Sequential()

model.add(Input(shape=(5,)))
model.add(Embedding(input_dim=10, output_dim=4))

model.add(Flatten())

model.add(Dense(1, activation='sigmoid'))

model.compile(optimizer='adam',loss='binary_crossentropy')

model.summary()

output = model.predict(input_data)
print("임베딩 레이어의 출력:\n", output)

Model: "sequential_2"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ embedding_2 (Embedding)              │ (None, 5, 4)                │              40 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ flatten_1 (Flatten)                  │ (None, 20)                  │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_1 (Dense)                      │ (None, 1)                   │              21 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 61 (244.00 B)

 Trainable params: 61 (244.00 B)

 Non-trainable params: 0 (0.00 B)

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 63ms/step
임베딩 레이어의 출력:
 [[0.5062941]
 [0.5062498]]


Q. imdb 영화 리뷰 데이터셋을 사용하여 긍부정 이진분류 모델링 및 평가를 수행

In [16]:
X_train.max()

9999

In [15]:
from keras.datasets import imdb
from keras.preprocessing.sequence import pad_sequences
from keras.models import Sequential
from keras.layers import Embedding, Flatten, Dense

# num_words 훈련 데이터에서 가장 자주 나타나는 상위 만개의 단어만 사용하겠다는 의미
(X_train, y_train),(X_test, y_test) = imdb.load_data(num_words=10000)

X_train = pad_sequences(X_train, maxlen=100)
X_test = pad_sequences(X_test, maxlen=100)

model = Sequential()
model.add(Input(shape=(100,))) # 모든 입력 데이터를 동일한 길이로 맞춰야 하기 때문에 패딩 사용 후 100으로 맞춤
model.add(Embedding(10000,8)) # 입력 차원은 단어 인덱스의 최대값 +1, 출력 차원은 임베딩 후 벡터 크기
# 임베딩 레이어는 모델이 학습할 수 있는 단어 인덱스의 총 개수를 설정
model.add(Flatten()) # 임베딩된 시퀀스를 평탄화
model.add(Dense(1, activation='sigmoid')) # 긍정과 부정을 분류하므로 시그모이드 활성화 함수 사용

model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
model.summary()

history = model.fit(X_train, y_train, epochs=10, batch_size=32, validation_split=0.2)

loss, accuracy = model.evaluate(X_test, y_test)
print(f"Test Accracy: {accuracy}")

17464789/17464789 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step


Model: "sequential_3"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ embedding_3 (Embedding)              │ (None, 100, 8)              │          80,000 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ flatten_2 (Flatten)                  │ (None, 800)                 │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_2 (Dense)                      │ (None, 1)                   │             801 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 80,801 (315.63 KB)

 Trainable params: 80,801 (315.63 KB)

 Non-trainable params: 0 (0.00 B)

Epoch 1/10
625/625 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.5930 - loss: 0.6644 - val_accuracy: 0.8278 - val_loss: 0.4132
Epoch 2/10
625/625 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8656 - loss: 0.3440 - val_accuracy: 0.8412 - val_loss: 0.3464
Epoch 3/10
625/625 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step - accuracy: 0.9127 - loss: 0.2401 - val_accuracy: 0.8538 - val_loss: 0.3328
Epoch 4/10
625/625 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - accuracy: 0.9439 - loss: 0.1768 - val_accuracy: 0.8462 - val_loss: 0.3437
Epoch 5/10
625/625 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.9634 - loss: 0.1305 - val_accuracy: 0.8520 - val_loss: 0.3575
Epoch 6/10
625/625 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.9813 - loss: 0.0913 - val_accuracy: 0.8456 - val_loss: 0.3810
Epoch 7/10
625/625 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step - accuracy: 0.9917 - loss: 0.0614 - val_accuracy: 0.8428 - val_loss: 0.4004
Epoch 8/10
625/625 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step - accuracy: 0.9962 - loss: 0.0433 - val_accuracy: 0.

Output Shape는 해당 레이어의 출력 텐서의 형태(Shape)를 나타냅니다. 모델에서 각 레이어가 입력을 받아서 어떤 형태로 출력을 생성하는지를 보여줍니다.

각 레이어의 Output Shape을 해석해보면:

Embedding Layer (embedding_3):

Output Shape: (None, 100, 8)
None: 배치 크기를 의미합니다. 훈련 중에는 다양한 크기의 배치를 사용할 수 있기 때문에 None으로 표시됩니다.
100: 입력된 시퀀스(리뷰)의 길이를 나타냅니다. 이 경우 길이 100의 단어 인덱스를 입력받습니다.
8: 각 단어 인덱스가 변환된 임베딩 벡터의 차원입니다. 즉, 각 단어가 8차원 벡터로 표현됩니다.
Flatten Layer (flatten_2):

Output Shape: (None, 800)
800은 100개의 단어가 각각 8차원 벡터로 변환되었기 때문에 계산된 값입니다:
100
×
8
=
800
100×8=800.
Flatten 레이어는 다차원 배열을 1차원 배열로 변환합니다.
Dense Layer (dense_2):

Output Shape: (None, 1)
1: 출력이 1개의 노드로, 이진 분류 문제이므로 긍정(1) 또는 부정(0) 중 하나의 값을 예측합니다.
2. Param
**Param #**는 해당 레이어에서 학습해야 하는 파라미터의 수를 나타냅니다. 파라미터는 모델이 학습하는 가중치와 바이어스 등을 포함합니다.

각 레이어의 파라미터 수를 해석해보면:

Embedding Layer (embedding_3):

Param #: 80,000
임베딩 레이어의 파라미터 수는
10
,
000
 (단어 수)
×
8
 (임베딩 차원)
=
80
,
000
10,000 (단어 수)×8 (임베딩 차원)=80,000입니다. 각 단어 인덱스에 대해 8차원의 가중치 벡터가 존재하므로 총 80,000개의 파라미터가 있습니다.
Flatten Layer (flatten_2):

Param #: 0
Flatten 레이어는 단순히 텐서의 차원을 변경하는 역할을 하므로, 학습할 파라미터가 없습니다.
Dense Layer (dense_2):

Param #: 801
이 레이어의 파라미터 수는 가중치와 바이어스를 포함합니다.
가중치 수는
800
 (입력 노드 수)
×
1
 (출력 노드 수)
=
800
800 (입력 노드 수)×1 (출력 노드 수)=800이며, 여기에 바이어스 1개가 추가되어 총 801개의 파라미터가 됩니다.

Output Shape: 각 레이어의 출력 형태를 나타냅니다.
Param #: 해당 레이어에서 학습해야 하는 파라미터의 수를 나타냅니다.

In [18]:
import numpy as np
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences

# 예측할 새로운 텍스트 데이터
new_reviews = [
    "This movie was fantastic! I really enjoyed it.",
    "Terrible movie, I will never watch it again.",
    "It was an average film, nothing special.",
    "Absolutely loved the movie, great acting and story!",
    "Worst movie ever, completely wasted my time."
]

# 이미 학습된 토크나이저로 시퀀스를 변환
word_index = imdb.get_word_index() # 함수는 IMDB 데이터셋에 사용된 단어와 그 단어에 대응하는 인덱스를 매핑한 딕셔너리를 반환
tokenizer = Tokenizer(num_words=10000)
# v + 3은 원래 IMDB 단어 인덱스에서 사용된 정수 인덱스 값에 3을 더한 새로운 인덱스를 생성
tokenizer.word_index = {k: (v + 3) for k, v in word_index.items()} # 특정 예약된 인덱스(예: <PAD>, <START>, <UNK> 등)를 위해 앞부분의 인덱스를 비워두기 위함
tokenizer.word_index["<PAD>"] = 0 # 패딩 토큰을 나타내며, 인덱스 0을 부여
tokenizer.word_index["<START>"] = 1 # 시퀀스의 시작을 나타내는 토큰이며, 인덱스 1을 부여
tokenizer.word_index["<UNK>"] = 2 # 인덱스에 포함되지 않은 단어들을 대체하기 위해 사용됩니다. 인덱스 2를 부여
tokenizer.word_index["<UNUSED>"] = 3 # 사용되지 않는 토큰을 위해 인덱스 3을 부여

sequences = tokenizer.texts_to_sequences(new_reviews)
padded_sequences = pad_sequences(sequences, maxlen=100)

# 예측 수행
predictions = model.predict(padded_sequences)

# 예측 결과 출력
for i, review in enumerate(new_reviews):
    print(f"Review: {review}")
    print(f"Predicted Sentiment: {'Positive' if predictions[i] > 0.5 else 'Negative'}\n")

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 21ms/step
Review: This movie was fantastic! I really enjoyed it.
Predicted Sentiment: Negative

Review: Terrible movie, I will never watch it again.
Predicted Sentiment: Negative

Review: It was an average film, nothing special.
Predicted Sentiment: Negative

Review: Absolutely loved the movie, great acting and story!
Predicted Sentiment: Negative

Review: Worst movie ever, completely wasted my time.
Predicted Sentiment: Negative

